add UnsupportedWemo

Spectre5 opened the pull request:
## Description:  Return an ~~UnknownDevice~~ UnsupportedWemo if pywemo doesn't recognize an expected Wemo device

New Wemo products won't yet exist in pywemo and currently discovery would return `None` for those devices.  This PR adds a mechanism to allow a user to manually interact with the product, assuming that it "looks" like other wemo products.  The idea for this was prompted by #177 and I'll re-post a few of those thoughts here.

If a devices does not exist in pywemo, then `device_from_uuid_and_location` returns `None`, so the device is never "discovered".  This PR would add an `UnknownDevice` which is basically the same as the base `Device` class.  This would allow some future Wemo device to still be discovered, just not directly usable since it wouldn't have the helper methods like `on`, `off`, etc defined for it.  But at least it would let a user know that something was discovered on the network.

I'm not sure if there would be downstream effects to this (i.e. in Home Assistant).  Downstream users could though easily ignore devices that return `True` for `isinstance(object, UnknownDevice)`.  Home Assistant may want to do this since the unknown device can't be easily controlled.

**Related issue (if applicable):**  N/A

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] There is no commented out code in this PR.

Spectre5 commented:
If we're worried about downstream issues due to this, we could also add a `debug` or `use_unknown_device` or similar keyword to `discover_devices` that gets passed to the `device_from_uuid_and_location` method.  Then give it a default of `False` and only if it is `True` would an `UnknownDevice` be returned.

pavoni commented on pywemo/discovery.py:
```diff
@@ -93,15 +93,27 @@ def device_from_uuid_and_location(uuid, mac, location,
+        try:
+            # unknown device, try to make a basic object to allow manual
+            # interaction
+            return UnknownDevice(url=location, mac=mac,
```
Should this also log something, along with some info about the device (eg `unsupported device X found - returning  Unknown Device`).

Spectre5 replied:
I've added a short log here, but see further discussion below.

pavoni commented on pywemo/discovery.py:
```diff
@@ -93,15 +93,27 @@ def device_from_uuid_and_location(uuid, mac, location,
+            # interaction
+            return UnknownDevice(url=location, mac=mac,
+                                 rediscovery_enabled=rediscovery_enabled)
+        except (AttributeError, requests.exceptions.RequestException):
```
Slightly concerned about this `failing silently`. I recall seeing an error about retrieving `Setup.XML` at some point. It might have been before we checked the device manufacturer - or might have been a crashed WeMo device.

Fine to catch this is you think it's needed - but perhaps put something in the log?

I think by the time we get here - we know it is a Belkin Device - but if that's what you want to check for - perhaps don't return an Unknown device.

Spectre5 replied:
see my new comment below, but yes we do know this is a Belkin device.  So calling it "unknown" is perhaps a bit of a misnomer, but I just meant that the product is unknown to pywemo.

I caught those two exceptions just because I didn't want some unfamiliar product to cause issues with the continued discovery - perhaps a future product will have a different setup.xml schema.  The `AttributeError` can be raised if a setup.xml file is found but doesn't contain the information needed (for example if it didn't include `friendlyName`).  But I'm happy to just let it raise an exception instead too.

pavoni reviewed (commented):
I think this is a great idea - a few small suggestions.

Spectre5 commented:
Let's talk about how to best implement this.  Based on what I said above and what you've noticed in the code, I think we need to resolve the following:

- Should there be a unique `UnknownDevice` returned or simple a base class `Device` object?  We do know it is a Belkin product already from the snippet below in the discovery method.  I though it would make sense to return an object that makes it clear that pywemo doesn't really know about the product, but I'm not set on that if we'd rather just return `Device`.  Maybe it would be more clear to return `UnknownWemo`, `ManualWemo`, or something else like that?

        if entry.match_device_description(
            {'manufacturer': 'Belkin International Inc.'}
        ):

- Should the new device (or basic `Device` class) object always be returned, or only with some sort of `debug` or similar argument set to `True`?  If so, what argument name would you think is most clear?  This probably depend on the point above.

pavoni commented:
Agree with you comments about `Unknown`, but I like your approach to split it out as a new class - because it makes it very clear to an API user that there is a device that we don't support. Perhaps just change the classname to `UnsupportedWemoDevice` or something similar?

As long as we log an error I'm fine with you catching those exceptions, but don't like to catch exceptions like that silently unless needed. Spent several days once once finding a bug where a library author was catching something he probably shouldn't have been - and doing it silently!

I see two classes of users we need to tell a device is unsupported. The API users are perfectly handled by your `UnknownDevice` class (whatever you decide to call it).  End users who have software that uses the API are probably best served by a clear error log - just so they aren't surprised that their device isn't supported.

In HA we'll probably trap the `UnknownDevice` and give an error anyway - but I still think it's good for the library to clearly report a warning / error.

Spectre5 commented:
Ok, I've removed the exception catching and renamed it to `UnsupportedWemo`.  Still happy to rename it again to whatever makes the most sense.  I also considered `UnsupportedDevice` since it could potentially not be a "Wemo" product...

This would raise an error if you call `device_from_uuid_and_location` directly with bad data, which is fine.  My only concern would be if an error/exception caused the rest of the discovery to be stopped.  Below are a few examples of failures that could happen, although these examples should not happen from the `discover_devices` function since the URL should be good and the the device should be a Belkin product.  I could see this potentially causing an issue if Belkin releases a product someday that doesn't follow the specification of the rest of the wemo products, which would probably result in an `AttributeError` (or something else?).  But we can probably cross that bridge when the day comes.

Also, these examples would be issues even if the uuid was a known product.  So these are not unique to the new class exactly, but it is a higher chance that a bad request is sent to it since we'd be fairly certain that the recognized wemo products would behave as expected.

- A bad request (capital XML)

    >>> pywemo.discovery.device_from_uuid_and_location('uuid:Unknown', None, r'http://192.168.1.30:49153/setup.XML')
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
      File "/home/scott/github/pywemo/pywemo/discovery.py", line 136, in device_from_uuid_and_location
        return UnsupportedWemo(
      File "/home/scott/github/pywemo/pywemo/ouimeaux_device/__init__.py", line 118, in __init__
        service_list = self._config.serviceList
    AttributeError: 'NoneType' object has no attribute 'serviceList'

- A bad port

    >>> pywemo.discovery.device_from_uuid_and_location('uuid:Unknown', None, r'http://192.168.1.30:50000/setup.xml')
    Traceback (most recent call last):
      File "/home/scott/github/pywemo/.venv/lib/python3.8/site-packages/urllib3/connection.py", line 169, in _new_conn
        conn = connection.create_connection(
      File "/home/scott/github/pywemo/.venv/lib/python3.8/site-packages/urllib3/util/connection.py", line 96, in create_connection
        raise err
      File "/home/scott/github/pywemo/.venv/lib/python3.8/site-packages/urllib3/util/connection.py", line 86, in create_connection
        sock.connect(sa)
    ConnectionRefusedError: [Errno 111] Connection refused
    ...<snip>...
    requests.exceptions.ConnectionError: HTTPConnectionPool(host='192.168.1.30', port=50000): Max retries exceeded with url: /setup.xml (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f05e4bd3070>: Failed to establish a new connection: [Errno 111] Connection refused'))

- A bad IP address results in a timeout ConnectionError, of course

Spectre5 commented:
For what it is worth, this is my one worry with not trying to catch exceptions, "That said, I feel like one bad wemo shouldn't bring down the whole platform?" from [this HA issue](https://github.com/home-assistant/core/issues/23744).  Another option might be to add a `failsafe` boolean to the discovery function that would log and ignore certain exceptions like `AttributeError`, xml exceptions, and maybe requests exceptions.  Don't want to make it too broad of course, as that is never a good thing either, but if it is trying to be `failsafe` (default false), then I assume that we'd want to be relatively broad.

The solution in that issue was to unplug and re-plugin the device, I don't think we can help with that!  And that might be a case where swallowing the exception and ignoring the device would be bad...

Spectre5 committed:
add UnsupportedDevice

pavoni commented:
Thanks for this - now merged and released as `0.5.6`

Do you want to do the HA update - or shall I?

Spectre5 commented:
If you have the time to do it, that'd be great.  But I'd like to see the PR you create (for my own understanding on the HA PR process).  If you can't do it right now, then I can take a stab at it in a couple days, probably.

pavoni commented:
HA PR here https://github.com/home-assistant/core/pull/44476
